src/library_dylink.js

```diff
@@ -896,6 +896,37 @@ var LibraryDylink = {
     return dso;
   },
 
+  $isValidFile__proxy: 'sync',
+  $isValidFile__deps: ['$FS'],
+  $isValidFile: function(filenamePtr) {
+    var filename = UTF8ToString(filenamePtr);
+#if DYLINK_DEBUG
+    dbg('isValidFile: ' + filename);
+#endif
+    var target = FS.findObject(filename);
+    return target && !target.isFolder && !target.isDevice;
+  },
+
+  $readFileProxied__proxy: 'sync',
+  $readFileProxied__deps: ['$FS', 'malloc'],
+  $readFileProxied: function(filenamePtr, outPtr) {
+    var filename = UTF8ToString(filenamePtr);
+#if DYLINK_DEBUG
+    dbg('readFileProxied: ' + filename);
+#endif
+    if (!FS.findObject(filename)) {
+      return 0;
+    }
+    var fileData = FS.readFile(filename, {encoding: 'binary'});
+    if (!(fileData instanceof Uint8Array)) {
+      fileData = new Uint8Array(fileData);
+    }
+    var mem = _malloc(fileData.length)
+    HEAP8.set(fileData, mem);
+    {{{ makeSetValue('outPtr', 0, 'mem', '*') }}};
+    return fileData.length;
+  },
```

Comment:
Could this be in C? Then it would just work with wasmFS. That is, I wonder if we can do more of the dlopen logic in C rather than JS.

Reply:
Perhaps, but I think that would require a larger refactor.   I'll add a comment here.

BTW, this code will also just work with wasmFS, even though the proxying is not necessary there perhaps.

BTW, with wasmfs I guess the JS API will work from any thread, unlike the old FS?  So this change would not have been needed at all there?

Reply:
Will the other PRs I saw you open on dlopen in native code not help towards this goal?

Reply:
I've now moved all the file I/O code into the native side and renamed this PR.  PTAL